File: src/Xlide.Vbe.Shim/Engine/HostApp.cs
namespace Xlide.Vbe.Shim.Engine;

/// <summary>
/// WHICH OFFICE APPLICATION THIS ADD-IN IS LOADED INTO.
///
/// The VBE is shared: the same add-in loads in Word, PowerPoint, Access and Outlook as readily as
/// in Excel, and until 2026-08-18 the language service assumed Excel in all of them - Word's
/// ThisDocument was told it was an Excel.Worksheet and offered a worksheet's members (the owner,
/// having run it in Word).
///
/// A NORMALISED WORD RATHER THAN THE PATH, because the engine should not be parsing executable
/// names to decide what it knows. The path is the shim's business; the answer is a token.
///
/// Read from the process rather than from the object model. `Application.Name` would need a
/// crossing and a host that is willing to answer during startup, and the executable already says
/// it: the add-in is loaded BY the host, so the process it is running in IS the host.
/// </summary>
internal static class HostApp
{
    /// <summary>
    /// `excel`, `word`, `powerpoint`, `access`, `outlook`, `visio`, `project`, or `other`.
    ///
    /// `other` is a real answer and the important one: it is what a host this product has never
    /// been run in returns, and the engine's rule for it is to assert nothing about document
    /// modules rather than to guess. Being silent in an unknown host beats being wrong in it.
    /// </summary>
    public static string Name { get; } = FromProcess();

    /// <summary>
    /// Whether this host's VBE carries the MSForms designer at all. Access's does not - Access
    /// VBA has its own Forms and no UserForms, its VBE offers no Insert > UserForm, and
    /// `VBComponents.Add(3)` can only fail there - so every surface that would CREATE a
    /// userform consults this and refuses plainly instead of relaying a COM error (the owner,
    /// 2026-08-19). Reading forms is not gated: a form component cannot exist in such a host.
    /// </summary>
    public static bool CarriesMsForms => Name is not "access";

    /// <summary>
    /// The name this host's <c>Application.Run</c> understands for a procedure, given the file
    /// the project belongs to (null when it has none yet).
    ///
    /// THREE HOSTS, THREE SPELLINGS, each measured against a live one rather than assumed:
    ///
    ///   Excel   <c>'Book1.xlsm'!Module.Proc</c>  Run resolves an unqualified name against the
    ///           ACTIVE workbook, so once two are open the name has to say which file it means,
    ///           or the host answers that the macro may not be available (2026-08-07).
    ///   Word    <c>Module.Proc</c>               Word's Run resolves across the open projects
    ///           and refuses Excel's form outright (2026-08-19, the day the Immediate window
    ///           first ran in Word).
    ///   Access  <c>Proc</c>                      Access has ONE database per application and
    ///           reads a dotted prefix as a LIBRARY DATABASE rather than a module, so both of
    ///           the other forms are refused. Measured 2026-09-06 against an open database:
    ///           `DiscountRate` answered 0.1, while `Pricing.DiscountRate` and the bang form
    ///           both came back "Microsoft Access cannot find the procedure". This is why the
    ///           test runner and the Immediate window could not run anything at all there.
    ///
    /// A host nobody has run this in keeps Excel's form, which is where every host started.
    /// The bare form is why a generated entry point has to carry a name of this product's own:
    /// in Access there is no module to disambiguate it with.
    /// </summary>
    public static string RunTarget(string? file, string module, string procedure) => Name switch
    {
        "access" => procedure,
        "word" => $"{module}.{procedure}",
        _ => file is null ? $"{module}.{procedure}" : $"'{file}'!{module}.{procedure}",
    };

    private static string FromProcess()
    {
        var image = Path.GetFileNameWithoutExtension(Environment.ProcessPath ?? string.Empty);
        return image.ToUpperInvariant() switch
        {
            "EXCEL" => "excel",
            "WINWORD" => "word",
            "POWERPNT" => "powerpoint",
            "MSACCESS" => "access",
            "OUTLOOK" => "outlook",
            "VISIO" => "visio",
            "WINPROJ" => "project",
            _ => "other",
        };
    }
}

File: engine/src/hostApp.ts
// WHICH OFFICE APPLICATION THIS ENGINE IS ANSWERING FOR, and the object model that answers
// with it.
//
// The VBE is shared - the add-in loads in Word, PowerPoint, Access and Outlook as readily as in
// Excel - and until 2026-08-18 every one of them was told it was Excel: Word's ThisDocument was
// typed as an Excel.Worksheet and offered a worksheet's members (the owner, having run it
// there). The host token arrived first; the models arrived with the analyzer's host registry
// (xlide_vscode 4.0.0, issues #24/#25), and this file is where the two meet.
//
// PROCESS STATE RATHER THAN A THREADED PARAMETER, and deliberately. The engine is a child of ONE
// Office process and the host cannot change under it, so this is a fact about the run rather than
// about a request. Threading it through assembleContext would put it in three call signatures -
// completion, hover, onType - to say the same unchanging thing at each of them.
//
// It arrives with `project/open`, which is the one message guaranteed to be sent before anything
// is asked, and defaults to Excel so an older shim against a newer engine behaves as it always
// did.

import {
    EMPTY_HOST_MODEL,
    hostObjectModelForToken,
} from '../../../xlide_vscode/src/analyzer/host/hostRegistry';
import { getExcelObjectModel, type HostObjectModel } from '../../../xlide_vscode/src/analyzer';

let current = 'excel';

/** Records the host, from `project/open`. An absent or empty value leaves the default alone. */
export function setHostApp(host: string | undefined): void {
    if (typeof host === 'string' && host.length > 0) {
        current = host.toLowerCase();
    }
}

export function hostApp(): string {
    return current;
}

/**
 * The registry's model for the current host, or undefined for Excel - the registry answers
 * undefined there ON PURPOSE, so every downstream `?? getExcelObjectModel()` default keeps the
 * behavior Excel always had. Callers that thread a model into the analyzer's resolvers want
 * exactly this shape: nothing to thread in Excel, the host's own model elsewhere, and the
 * empty model (every lookup misses, nothing asserted) in a host the registry does not know.
 */
export function currentHostModelOverride(): HostObjectModel | undefined {
    return hostObjectModelForToken(current);
}

/** The current host's model, Excel's included: for callers that read it rather than thread it. */
export function currentHostModel(): HostObjectModel {
    return currentHostModelOverride() ?? getExcelObjectModel();
}

/**
 * Whether the host's own object model is one this engine can speak about. Excel, Word,
 * PowerPoint and Access today - whatever the registry holds, rather than a list kept here.
 * A host it answers with the EMPTY model for (Outlook, Visio, unknowns) is honestly unknown:
 * document modules assert no type and the knowledge routes say so, which beats being
 * confidently wrong in someone else's application.
 */
export function hostModelIsKnown(): boolean {
    return currentHostModel() !== EMPTY_HOST_MODEL;
}

/**
 * Whether the current host's VBE carries the MSForms designer at all. Access's does not -
 * Access VBA has its own Forms and no UserForms, so its VBE offers no Insert > UserForm and
 * `VBComponents.Import` of a .frm can only fail there (the owner, 2026-08-19). Everything that
 * would CREATE a userform consults this; reading one is not gated, because a form component
 * can never exist in such a host to begin with.
 */
export function hostCarriesMsForms(): boolean {
    return current !== 'access';
}
